crates/symbolicator-service/src/services/symbolication/sourcemap.rs

```
@@ -0,0 +1,451 @@
+                    },
+                    filename: Some(token.name().unwrap_or_default().to_string()),
+                    abs_path: token.file_name().unwrap_or_default().to_string(),
+                    lineno: Some(token.line()),
+                    colno: Some(token.column()),
+                    pre_context,
+                    context_line: Some(context_line.to_string()),
+                    post_context,
+                    ..frame.clone()
+                },
+            };
+
+            Ok(result)
+        } else {
+            Err(FrameStatus::Missing)
+        }
+    } else {
+        Err(FrameStatus::Missing)
+    }
+}
+
+// Transforms a full absolute url into 2 or 4 generalized options. Based on `ReleaseFile.normalize`.
+// https://github.com/getsentry/sentry/blob/master/src/sentry/models/releasefile.py
+fn get_release_file_candidate_urls(url: &str) -> Vec<String> {
+    let mut urls = vec![];
+
+    if let Ok(url) = Url::parse(url) {
+        // Absolute without fragment
+        urls.push(url[..Position::AfterQuery].to_string());
+
+        // Absolute without query
+        if url.query().is_some() {
+            urls.push(url[..Position::AfterPath].to_string())
+        }
+
+        // Relative without fragment
+        urls.push(format!(
+            "~{}",
+            &url[Position::BeforePath..Position::AfterQuery]
+        ));
+
+        // Relative without query
+        if url.query().is_some() {
+            urls.push(format!(
+                "~{}",
+                &url[Position::BeforePath..Position::AfterPath]
+            ));
+        }
+    }
+
+    urls
+}
+
+// Joins together frames `abs_path` and discovered sourcemap reference.
+fn resolve_sourcemap_url(abs_path: &str, sourcemap_location: &str) -> anyhow::Result<String> {
+    let base = Url::parse(abs_path)?;
+    base.join(sourcemap_location)
+        .map(|url| url.to_string())
+        .map_err(|e| e.into())
+}
```

I would suggest combining these two functions into one.

1. `resolve_sourcemap_url` is only called immediately before `get_release_file_candidate_urls`.
1. The error returned by `resolve_sourcemap_url` is never handled, it just gets silently discarded by `if let Ok(…)`.
2. The functions interface via a `String` when they are really both dealing with `Url`s.